src/replication.c

```diff
@@ -1925,23 +1948,24 @@ replicate(ISTGT_LU_DISK *spec, ISTGT_LU_CMD_Ptr cmd, uint64_t offset, uint64_t n
 		// check for status of rcomm_cmd
 		rc = check_for_command_completion(spec, rcomm_cmd, cmd);
 		if (rc) {
-			if (rc == 1)
+			if (rc == 1) {
 				rc = cmd->data_len = rcomm_cmd->data_len;
+			}
 			rcomm_cmd->state = CMD_EXECUTION_DONE;
-			put_to_mempool(&spec->rcommon_deadlist, rcomm_cmd);
-			MTX_UNLOCK(rcomm_cmd->mutex);
 
 			/*
 			 * NOTE: This is for debugging purpose only
 			 */
 			if (err_num == ETIMEDOUT)
-				fprintf(stderr,"last errno(%d) opcode(%d)\n",
-				    errno, rcomm_cmd->opcode);
+				fprintf(stderr, "last errno(%d) "
+				    "opcode(%d)\n", errno, rcomm_cmd->opcode);
 
 			MTX_LOCK(&spec->rq_mtx);
 			TAILQ_REMOVE(&spec->rcommon_waitq, rcomm_cmd, wait_cmd_next);
 			MTX_UNLOCK(&spec->rq_mtx);
 
+			put_to_mempool(&spec->rcommon_deadlist, rcomm_cmd);
+			MTX_UNLOCK(rcomm_cmd->mutex);
```

Comment:
can this lock sequence lead to deadlock?
And, why do we need lock to add to mempool?